File: multi-language/julia/ucd.jl
# multi-language/julia/ucd.jl
using Statistics

function verify_conservation(C, F; tol=1e-10)
    return abs(C + F - 1.0) < tol
end

struct UCD
    data::Matrix{Float64}
end

function analyze(ucd::UCD)
    n = size(ucd.data, 1)
    if n > 1
        corr_mat = abs.(cor(ucd.data'))
        C = (sum(corr_mat) - n) / (n * (n - 1))
    else
        C = 0.5
    end
    F = 1.0 - C
    return (C=C, F=F, conservation=verify_conservation(C, F))
end

# Example
data = [1.0 2.0 3.0 4.0; 2.0 3.0 4.0 5.0; 5.0 6.0 7.0 8.0]
ucd = UCD(data)
println(analyze(ucd))

File: multi-language/r/ucd.R
# multi-language/r/ucd.R
ucd_analyze <- function(data) {
  if (nrow(data) > 1) {
    corr_mat <- abs(cor(t(data)))
    C <- (sum(corr_mat) - nrow(data)) / (nrow(data) * (nrow(data) - 1))
  } else {
    C <- 0.5
  }
  F <- 1.0 - C
  return(list(C = C, F = F, conservation = (abs(C + F - 1.0) < 1e-10)))
}

data <- matrix(c(1,2,3,4, 2,3,4,5, 5,6,7,8), nrow=3, byrow=TRUE)
print(ucd_analyze(data))
